rclc/src/rclc/action_goal_handle.c

```diff
@@ -0,0 +1,283 @@
+
+
+typedef struct rclc_generic_entity_t
+{
+  DECLARE_GOAL_HANDLE_POOL
+} rclc_generic_entity_t;
+
+
+void rclc_action_put_goal_handle_in_list(
+  rclc_action_goal_handle_t ** list,
+  rclc_action_goal_handle_t * goal_handle)
+{
+  goal_handle->next = *list;
+  *list = goal_handle;
+}
+
+bool rclc_action_check_handle_in_list(
+  rclc_action_goal_handle_t ** list,
+  rclc_action_goal_handle_t * goal_handle)
+{
+  rclc_action_goal_handle_t * handle = *list;
+  while (NULL != handle) {
+    if (handle == goal_handle) {
+      return true;
+    }
+    handle = handle->next;
+  }
+  return false;
+}
+
+rclc_action_goal_handle_t * rclc_action_pop_goal_handle_from_list(
+  rclc_action_goal_handle_t ** list)
+{
+  rclc_action_goal_handle_t * handle = *list;
+  *list = (*list == NULL ) ? NULL : (*list)->next;
+  return handle;
+}
+
+rclc_action_goal_handle_t * rclc_action_take_goal_handle_from_list(
+  rclc_action_goal_handle_t ** list,
+  rclc_action_goal_handle_t * goal_handle)
+{
+  rclc_action_goal_handle_t * handle = *list;
+  if (goal_handle == handle) {
+    *list = handle->next;
+    return goal_handle;
+  }
+
+  while (NULL != handle) {
+    if (handle->next == goal_handle) {
+      handle->next = goal_handle->next;
+      return goal_handle;
+    }
+    handle = handle->next;
+  }
+  return NULL;
+}
+
+rclc_action_goal_handle_t * rclc_action_take_goal_handle(
+  void * untyped_entity)
+{
+  rclc_generic_entity_t * entity = (rclc_generic_entity_t *) untyped_entity;
+  rclc_action_goal_handle_t * handle = rclc_action_pop_goal_handle_from_list(
+    &entity->free_goal_handles);
+
+  if (NULL != handle) {
+    // Initialize handle
+    handle->available_goal_response = false;
+    handle->goal_accepted = false;
+    handle->available_feedback = false;
+    handle->available_result_response = false;
+    handle->available_cancel_response = false;
+    handle->goal_cancelled = false;
+    handle->status = GOAL_STATE_UNKNOWN;
+    rclc_action_put_goal_handle_in_list(&entity->used_goal_handles, handle);
+  }
+
+  return handle;
+}
+
+void rclc_action_init_goal_handle_memory(
+  void * untyped_entity)
+{
+  rclc_generic_entity_t * entity = (rclc_generic_entity_t *) untyped_entity;
+  entity->free_goal_handles = entity->goal_handles_memory;
+  size_t size = entity->goal_handles_memory_size;
+  for (size_t i = 0; i < size - 1; i++) {
+    entity->free_goal_handles[i].next = &entity->free_goal_handles[i + 1];
+  }
+  entity->free_goal_handles[size - 1].next = NULL;
+}
+
+void rclc_action_put_goal_handle(
+  void * untyped_entity,
+  rclc_action_goal_handle_t * goal_handle)
+{
+  rclc_generic_entity_t * entity = (rclc_generic_entity_t *) untyped_entity;
+  rclc_action_goal_handle_t * handle = rclc_action_take_goal_handle_from_list(
+    &entity->used_goal_handles, goal_handle);
+  if (NULL != handle) {
+    rclc_action_put_goal_handle_in_list(&entity->free_goal_handles, handle);
+  }
+}
+
+rclc_action_goal_handle_t * rclc_action_get_handle_by_uuid(
```

Comment:
Rename to `rclc_action_get_goal_handle_by_uuid`, i.e. include `goal`?

Reply:
d2e52f92a7286287271489b6407c28b8aee4ca0b
4e02ee12e55098437304091a3f37a1c60aa501a4